File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        def dfs(root,mn,mx)->int:
            if (root==None) :
                return mx - mn
            mx = max(mx, root.val)
            mn = min(mn, root.val)
            return max(dfs(root.left, mn, mx), dfs(root.right, mn, mx))
        return dfs(root,root.val,root.val)
        

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.java

class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        return dfs(root,root->val,root->val);
    }
    int dfs(TreeNode* root,int mini,int maxi)
    {
        if(root==NULL) return maxi-mini;
        maxi=max(root->val,maxi);
        mini=min(root->val,mini);
        return max(dfs(root->left,mini,maxi),dfs(root->right,mini,maxi));
        
    }
};